### DISCUSSION TITLE
URQL One-off Queries and Mutations + Cache Management Issue

### DISCUSSION THREAD

**Discussion by mbushpilot2b** [urql-graphql/urql#2777] [Q&A] [answered]
# Question
With [One-off Queries and Mutations](https://formidable.com/open-source/urql/docs/basics/core/#one-off-queries-and-mutations) how can I manage cache using `cache.invalidate()` in my `cacheExchange` and yet still receive the response back in our `.then(res)` of the `client.query()` or `client.mutation()`? 

## Expected
`cache.invalidate()` should not nullify the response to the `client.mutation().then(res => console.log())`

## Actual Result
`cache.invalidate()` runs first causing the `.then(res)` to return **`null`** every time

# Code Example
## cacheExchange
```
const newClient = createClient({
  exchanges: [
    cacheExchange<GraphCacheConfig>({
      updates: {
        Mutation: {
          updateWorklist(result, _args, cache, _info) {
            console.log("updateWorklistCache", result) // result is defined here {updateWorklist: { id: 24, name: "test", visibility: "PRIVATE", filters: {…}, order: null } … }
            if (result?.updateWorklist?.id) {
              cache.invalidate({ __typename: 'Worklist', id: result?.updateWorklist?.id})
            }
          },
        },
      },
    }),
  ]
})
```

## client.mutation()
```
const updateWorklist = (id: string, worklist: any) =>
client
  .mutation(UpdateWorklistDocument, { id, worklist })
  .toPromise()
  .then((res) => {
    console.log('res', res) // returns {data: {updateWorklist: null}}
    message.success('Worklist saved')
    return {
      id: res.data.updateWorklist.id,
      name: res.data.updateWorklist.name,
      visibility: res.data.updateWorklist.visibility,
      filter: res.data.updateWorklist.filters?.filters,
      sort: res.data.updateWorklist.filters?.order,
    }
  })
  .catch((_err: any) => message.error('Error saving worklist'))
```

@kitten This is what I emailed you about today. Thank you so much for all you have done for the urql project 🚀

**Comment by JoviDeCroock** [comment, marked as answer]
The question here doesn't really make sense to me 😅 calling `cache.invalidate` should have no impact on the result of your mutation i.e. your `res.data` is untouched however, your next `client.query` won't hit your cache and rather have to fetch.

However if you want to update your list you don't have to do anything that will happen automatically and doesn't need invalidation, so after that mutation your next query will be the new worklist.

Getting the response for your worklist is intended as your network response is relevant here and invalidating only affects the cache 😅

**Reply by mbushpilot2b** [reply]
Yeah so do you think this is a bug then? Because it certainly does affect the outcome. In fact the `console.log("updateWorklistCache", result)` in the `cacheExchange` prints **"before"** `console.log('res', res)` in the `client.mutation()`

**Reply by JoviDeCroock** [reply]
That is exactly the purpose of this, the updater runs as part of the exchange chain so no it's not a bug. You are saying that it removes the mutation response but I am currently not seeing that behavior when trying so myself 😅

In fact this is the exepcted flow:

- mutation runs in fetchExchange
- mutation response gets handled in cacheExchange (and hence updater)
- mutation response is returned to the invoker
- dependent queries are re-run as part of what was updated in your updater

So the call-order is def not a bug

**Reply by mbushpilot2b** [reply]
> ... if you want to update your list you don't have to do anything...

Well this is news to me, so I could swear I saw a stale version of my `worklists` list when updating a `worklist`. I will look into this.

> Getting the response for your worklist is intended as your network response is relevant here and invalidating only affects the cache 😅

Yes getting the response back IS what I expect when using cache.invalidate() but I am just getting `null`. I updated my initial post above to show what the responses are.

**Reply by mbushpilot2b** [reply]
> In fact this is the exepcted flow:
> - mutation runs in fetchExchange
> - mutation response gets handled in cacheExchange (and hence updater)
> - mutation response is returned to the invoker
> - dependent queries are re-run as part of what was updated in your updater

Yes this is the call order as you listed, and no thats not a bug. What does seem like a bug is: `mutation response is returned to the invoker` happens, but it is only returning `null` in this step.

**Reply by JoviDeCroock** [reply]
Yes, no, I can confirm that the response gets lost. Made a reproduction to help us talk about a common subject https://codesandbox.io/s/goofy-feather-wtz7w1?file=/src/index.js 😅 However much I do think it's uncommon to invalidate a response that is meant to update the cache, this needs to get fixed.

**Reply by mbushpilot2b** [reply]
I edited my question to remove the "After" bit. However, the issue where I receive `null` still exists. 

> if you want to update your list you don't have to do anything

I have confirmed this to be the case! Thank you for pointing that out! So i do not need to use cache.invalidate() but it is strange that with cache.invalidate() my response is null to the invoker?

**Reply by mbushpilot2b** [reply]
@JoviDeCroock Sweet, wow thank you for reproducing it in isolation, yes thats exactly the issue there. And yes I didn't realize it would be updated automatically, and yes, thats true it updates automatically without the need to invalidate but its still a bug that it returns null

**Reply by mbushpilot2b** [reply]
@JoviDeCroock Updating comes for free but deleting and creating do not. And it makes sense why it would not, who knows what the server logic is (sorting/filtering/searching) applied to a list of worklists in this case. So in these cases the only way i can handle this properly is to invalidate the cache. Since when doing this i get null, i have to revert to using the input I sent to the server during the mutation instead of the response, which is very bad practice, as the server could have business logic to do something else entirely. Can we make this an issue?
